x/posts/internal/keeper/invariants.go

```diff
@@ -0,0 +1,130 @@
+
+// Hash256PostIDInvariant checks that the all posts have a SHA256 ID
+func Hash256PostIDInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidPostIDs types.PostIDs
+		k.IteratePosts(ctx, func(_ int64, post types.Post) (stop bool) {
+			if !post.PostID.Valid() {
+				invalidPostIDs = append(invalidPostIDs, post.PostID)
+			}
+			return false
+		})
+
+		return sdk.FormatInvariant(types.ModuleName, "invalid post IDs",
+			fmt.Sprintf("The following post IDs are invalid:\n %s", formatOutputIDs(invalidPostIDs)),
+		), invalidPostIDs != nil
+	}
+}
+
+// ValidCommentsDateInvariants checks that comments creation date is always greater than parent creation date
+func ValidCommentsDateInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidCommentsIDs types.PostIDs
+		k.IteratePosts(ctx, func(_ int64, post types.Post) (stop bool) {
+			if post.ParentID.Valid() {
+				parentPost, _ := k.GetPost(ctx, post.ParentID)
+				if post.Created.Before(parentPost.Created) {
+					invalidCommentsIDs = append(invalidCommentsIDs, post.PostID)
+				}
+			}
+			return false
+		})
+
+		return sdk.FormatInvariant(types.ModuleName, "comments dates before parent post date",
+			fmt.Sprintf("The following post IDs referred to posts which are invalid comments "+
+				"having creation date before parent post creation date:\n %s",
+				formatOutputIDs(invalidCommentsIDs)),
+		), invalidCommentsIDs != nil
+	}
+}
+
+// ValidPostForReactionsInvariants checks that the post related to the reactions is valid and exists
+func ValidPostForReactionsInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidPostIDs types.PostIDs
+		reactions := k.GetReactions(ctx)
+		for key := range reactions {
+			postID := types.PostID(key)
+			if _, found := k.GetPost(ctx, postID); !found {
+				invalidPostIDs = append(invalidPostIDs, postID)
+			}
+		}
+
+		return sdk.FormatInvariant(types.ModuleName, "posts reactions refers to non existing posts",
+			fmt.Sprintf("The following post IDs referred to posts that did not exist:\n %s",
+				formatOutputIDs(invalidPostIDs)),
+		), invalidPostIDs != nil
+	}
+}
+
+// ValidPollForPollAnswersInvariants check that the poll answers are referred to a valid post's poll
+func ValidPollForPollAnswersInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidPostIDs types.PostIDs
+		answers := k.GetPollAnswersMap(ctx)
+		for key := range answers {
+			postID := types.PostID(key)
+			if post, found := k.GetPost(ctx, postID); !found || (found && post.PollData == nil) {
+				invalidPostIDs = append(invalidPostIDs, postID)
+			}
+		}
+
+		return sdk.FormatInvariant(types.ModuleName, "poll answers refers to posts without poll",
+			fmt.Sprintf("The following post IDs referred to posts that either not exists or has no poll associated:\n %s",
+				formatOutputIDs(invalidPostIDs)),
+		), invalidPostIDs != nil
```

Comment:
As before, we should probably have a list of answers that refer to non existing posts instead of a list of `PostID`. Also, we can improve the message telling "The following answers refer to a post that either does not exist or has no poll associated to it".